client/src/test/java/io/pravega/client/stream/impl/StreamSegmentsTest.java

```diff
@@ -144,4 +145,127 @@ public void testSameRoutingKey() {
+        assertEquals(getSegment(3, 1), streamSegments.getSegmentForKey(0.4));
+        assertEquals(getSegment(1, 0), streamSegments.getSegmentForKey(0.7));
+
+        // simulate fetch successors for segment 3.
+        newRange = new HashMap<>();
+        newRange.put(new SegmentWithRange(getSegment(8, 4), 0.25, 0.75),
+                ImmutableList.of(computeSegmentId(3, 1), computeSegmentId(4, 2)));
+        streamSegments = streamSegments.withReplacementRange(getSegment(3, 1), new StreamSegmentsWithPredecessors(newRange, ""));
+
+        assertEquals(getSegment(6, 3), streamSegments.getSegmentForKey(0.1));
+        assertEquals(getSegment(7, 3), streamSegments.getSegmentForKey(0.24));
+        assertEquals(getSegment(8, 4), streamSegments.getSegmentForKey(0.4));
+        //this should not return segment 8 as segment 4 is never updated.
+        assertEquals(getSegment(1, 0), streamSegments.getSegmentForKey(0.7));
+
+        // simulate fetch successors for segment 1.
+        newRange = new HashMap<>();
+        newRange.put(new SegmentWithRange(getSegment(4, 2), 0.5, 0.75),
+                ImmutableList.of(computeSegmentId(1, 0)));
+        newRange.put(new SegmentWithRange(getSegment(5, 2), 0.75, 1.0),
+                ImmutableList.of(computeSegmentId(1, 0)));
+        streamSegments = streamSegments.withReplacementRange(getSegment(1, 0), new StreamSegmentsWithPredecessors(newRange, ""));
+
+        assertEquals(getSegment(6, 3), streamSegments.getSegmentForKey(0.1));
+        assertEquals(getSegment(7, 3), streamSegments.getSegmentForKey(0.24));
+        // this should now return segment 8 as before
+        assertEquals(getSegment(8, 4), streamSegments.getSegmentForKey(0.4));
+        //this should now return segment 4 and not segment 1 or segment 8
+        assertEquals(getSegment(4, 2), streamSegments.getSegmentForKey(0.7));
+        //this should now return segment 5 and not segment 1.
+        assertEquals(getSegment(5, 2), streamSegments.getSegmentForKey(0.9));
+
+        // simulate fetch successors for segment 4.
+        newRange = new HashMap<>();
+        newRange.put(new SegmentWithRange(getSegment(8, 4), 0.25, 0.75),
+                ImmutableList.of(computeSegmentId(3, 1), computeSegmentId(4, 2)));
+        streamSegments = streamSegments.withReplacementRange(getSegment(4, 2), new StreamSegmentsWithPredecessors(newRange, ""));
+
+        assertEquals(getSegment(6, 3), streamSegments.getSegmentForKey(0.1));
+        assertEquals(getSegment(7, 3), streamSegments.getSegmentForKey(0.24));
+        assertEquals(getSegment(8, 4), streamSegments.getSegmentForKey(0.4));
+        // this should now return segment 8 and not segment 4
+        assertEquals(getSegment(8, 4), streamSegments.getSegmentForKey(0.7));
+        assertEquals(getSegment(5, 2), streamSegments.getSegmentForKey(0.9));
+
+    }
+
+    @Test
+    public void testRangeReplacementMultipleSegmentMerge() {
```

Comment:
Let's also test the incremental version of this, where the seal is encountered for 2 then for 0 and then for 1. As this is awkward to deal with as the same segment ends up briefly with non-contiguous keyspace.

Reply:
added a test case.